### DISCUSSION TITLE
TerminusDB and application development workflow

### DISCUSSION THREAD

**Discussion by NickDarvey** [terminusdb/terminusdb#1039] [Q&A]
I'm building an application for a domain where there's lots of navigating relationships and formal revisioning is important. I feel TerminusDB might be a good fit for persistence.

I want to learn more about how an application development lifecycle might include TerminusDB. I'll present a couple of scenarios and describe my best guess at how they should be approached based on what I know of TerminusDB. (Little!) I'd like to hear how you'd approach these scenarios.

## Changes to schema

If I were to be developing an application with a RDMS for persistence, I might deliver changes to schema using a [state-based](https://docs.microsoft.com/en-us/sql/ssdt/sql-server-data-tools?view=sql-server-ver15) or [migration-based](https://sqitch.org/docs/manual/sqitchtutorial/) approach. Practically this means I have my database schema defined (declaratively or procedurally, respective to the approach) in my Git repository alongside my application code. A new feature may involve a change to the database schema and application code. Delivering that change to production means getting the change merged in to the main branch at which point a continuous delivery pipeline will update the database's schema and update the application.

If I am to develop an application with TerminusDB for persistence, how might I use a similar approach or how does my thinking need to change? 

**Similar approach**
TerminusDB JSON schema documents can be stored in a Git repo alongside my application code, like a state-based approach with RDMSs. The continuous delivery pipeline can deliver these changes by using the replace schema API and updating the database's `local/branch/main` schema graph.

**Different approach**
The application development workflow shouldn't be the same when using TerminusDB. The database shouldn't be treated like 'just' a persistence layer for an application, where the application drives the schema. (Applications come and go, data is forever.) Instead it should be changed separately, using the branching and versioning provided by TerminusDB. The application depends on the tip of the main branch.

## Changes to data

Some changes to schema also demand changes to data. For example, we may start with a `FirstName` and `LastName` column and later decide that just a `FullName` column is more sensible. The schema migration here is introducing a new column (and later, removing the old ones), the data migration is concatenating the `FirstName` and `LastName` values to produce the initial value for the new `FullName` column. A state-based approach falters here and I need to [supplement it](https://github.com/4tecture/SSDTDataMigration) with migration-based changes to data.

How might I migrate data with TerminusDB?

**Approach**
TerminusDB doesn't have functionality that helps with data migrations when there are such changes to schema. You can make these kinds of changes by writing migration scripts with WOQL queries.

**Comment by GavinMendelGleason** [comment]
Really excellent questions @NickDarvey ! Let me try to do my best in suggesting approaches. 

# Schema Migration

TerminusDB stores schemata with their instance data in a single `Commit` object. This object stays in the history, so that you can recover the exact schema and data. You can dump only the schema JSON documents, or only the instance JSON documents easily at any time. 

Both can be migrated, as long as one does so in lock-step. That is, you can update the schema to a new schema, as long as you migrate the data along with you to reflect this, and the entire move happens in a single transaction.  If you are adding optional properties, this is fairly trivial. However, adding required properties or dropping existing properties makes things a little tricky. 

Unfortunately in the current version that is a little awkward through the document interface. The reason is that our documents are described by the previous schema state. 

The work around we employ at the minute is to temporarily turn-off schema checking, update the schema to an inconsistent schema, update the documents according to this new schema state and turn the schema checking back on. 

NOTE: It might be worth adding a flag to document updates that turns the schema back on for a given document update so that you can ensure that your data migration fixes any inconsistencies which might have appeared without changing the state if it does not. I'll discuss this with the engineering team as a possible feature request. 

Alternatively, if you can express your update as a WOQL query, you can do the update in a single step in WOQL (which can update the schema and instance data simultaneously).
